From c18445f900ffc5600c45c7c7274aacee64320b21 Mon Sep 17 00:00:00 2001
From: Rob Crittenden <rcritten@redhat.com>
Date: Tue, 5 Aug 2025 17:38:04 +0000
Subject: [PATCH 2/6] Add initial ML-DSA support with OpenSSL 3.5.0

OpenSSL 3.5.0 has added support for Post-Quantum ML-DSA keys.
Allow them to be generated by certmonger.

The minimum key size is not fully implemented because key size
has a different meaning in this context. It is plumbed in, such that
one can set a numeric size, but it isn't used and is effectively
meaningless.
---
 configure.ac                         |  50 +++++++++
 src/certmonger.conf.5.in             |   1 +
 src/getcert.c                        |  11 ++
 src/keygen-n.c                       |  22 ++++
 src/keygen-o.c                       |  30 +++++-
 src/keyiread-n.c                     |   2 +
 src/keyiread-o.c                     |  56 +++++++++-
 src/keyiread.c                       |  31 ++++++
 src/prefs.c                          |  14 +++
 src/store-files.c                    | 147 +++++++++++++++++++++++++++
 src/store-int.h                      |   5 +
 src/store.h                          |   1 +
 src/submit-e.c                       |  11 ++
 src/tdbush.c                         |  79 ++++++++++----
 tests/002-keygen-ml-dsa/expected.out |  14 +++
 tests/002-keygen-ml-dsa/run.sh       |  43 ++++++++
 tests/028-dbus/expected.out          |   8 +-
 tests/tools/keyiread.c               |  17 ++++
 18 files changed, 512 insertions(+), 30 deletions(-)
 create mode 100644 tests/002-keygen-ml-dsa/expected.out
 create mode 100755 tests/002-keygen-ml-dsa/run.sh

diff --git a/configure.ac b/configure.ac
index f621a07d..78e7d54c 100644
--- a/configure.ac
+++ b/configure.ac
@@ -712,6 +712,54 @@ if ! ${configure_dist_target_only:-false} ; then
 	AC_SUBST(MAN_EC)
 	AC_SUBST(NO_MAN_EC)
 
+	AC_ARG_ENABLE(ml_dsa,
+	AS_HELP_STRING(--disable-ml-dsa,[Disable ML-DSA key support, even if available.]),
+	ml_dsa=$enableval,
+	ml_dsa=maybe)
+	if test x$ml_dsa != xno ; then
+		CFLAGSsave="$CFLAGS"
+		LIBSsave="$LIBS"
+		CFLAGS="$OPENSSL_CFLAGS $CFLAGS"
+		LIBS="$OPENSSL_LIBS $LIBS"
+		AC_CHECK_LIB(crypto,EVP_PKEY_Q_keygen)
+		# CFLAGS="$NSS_CFLAGS $CFLAGS"
+		# LIBS="$NSS_LIBS $LIBS"
+		# AC_CHECK_FUNCS(SECKEY_CreateECPrivateKey)
+		# AC_CHECK_FUNCS(PK11_PrivDecrypt)
+		# AC_CHECK_FUNCS(NSS_OptionGet NSS_OptionSet)
+		CFLAGS="$CFLAGSsave"
+		LIBS="$LIBSsave"
+		can_ml_dsa=true
+		if test x$ac_cv_lib_crypto_EVP_PKEY_Q_keygen = xno ; then
+			can_ml_dsa=false
+		fi
+		AC_MSG_CHECKING([for ML-DSA support])
+		if $can_ml_dsa ; then
+			AC_MSG_RESULT(enabling ML-DSA support)
+			AC_DEFINE_UNQUOTED(CM_ENABLE_ML_DSA,1,[Define to enable ML-DSA support.])
+			MAN_ML_DSA=""
+			NO_MAN_ML_DSA=".\\\" "
+			ml_dsa=yes
+		else
+			if test x$ml_dsa != xyes ; then
+				AC_MSG_WARN([unavailable, disabling])
+				ml_dsa=no
+			else
+				AC_MSG_ERROR([unavailable])
+			fi
+			MAN_ML_DSA=".\\\" "
+			NO_MAN_EC=""
+		fi
+	else
+		AC_MSG_RESULT(NOT enabling ML-DSA support)
+		MAN_ML_DSA=".\\\" "
+		NO_MAN_ML_DSA=""
+		ml_dsa=no
+	fi
+	AM_CONDITIONAL(HAVE_ML_DSA,[test x$ml_dsa = xyes])
+	AC_SUBST(MAN_ML_DSA)
+	AC_SUBST(NO_MAN_ML_DSA)
+
 	AC_DEFINE_UNQUOTED(CM_DEFAULT_KEY_STORAGE_TYPE,cm_key_storage_nssdb,[Define to the default type of storage used for keys.])
 	AC_DEFINE_UNQUOTED(CM_DEFAULT_KEY_STORAGE_LOCATION,"/etc/pki/nssdb",[Define to the default location of storage used for keys.])
 	AC_DEFINE_UNQUOTED(CM_DEFAULT_KEY_TOKEN,NULL,[Define to the default token used for holding keys.])
@@ -730,9 +778,11 @@ if ! ${configure_dist_target_only:-false} ; then
 	AC_SUBST(CM_MINIMUM_RSA_KEY_SIZE)
 	CM_MINIMUM_DSA_KEY_SIZE=512
 	CM_MINIMUM_EC_KEY_SIZE=256
+	CM_MINIMUM_ML_DSA_KEY_SIZE=44
 	AC_DEFINE_UNQUOTED(CM_MINIMUM_RSA_KEY_SIZE,$CM_MINIMUM_RSA_KEY_SIZE,[Define to the minimum key size when generating RSA keys.  Requests to generate smaller keys will be forced to this key size.])
 	AC_DEFINE_UNQUOTED(CM_MINIMUM_DSA_KEY_SIZE,$CM_MINIMUM_DSA_KEY_SIZE,[Define to the minimum key size when generating DSA parameters and keys.  Requests to generate smaller keys will be forced to this key size.])
 	AC_DEFINE_UNQUOTED(CM_MINIMUM_EC_KEY_SIZE,$CM_MINIMUM_EC_KEY_SIZE,[Define to the minimum key size when selecting elliptic curve parameters.  Requests to generate smaller keys will be forced to this key size.])
+	AC_DEFINE_UNQUOTED(CM_MINIMUM_ML_DSA_KEY_SIZE,$CM_MINIMUM_ML_DSA_KEY_SIZE,[Define to the minimum key size when selecting ML-DSA parameters.  Requests to generate smaller keys will be forced to this key size.])
 	AC_SUBST(CM_MINIMUM_RSA_KEY_SIZE)
 	AC_SUBST(CM_MINIMUM_DSA_KEY_SIZE)
 	AC_SUBST(CM_MINIMUM_EC_KEY_SIZE)
diff --git a/src/certmonger.conf.5.in b/src/certmonger.conf.5.in
index 49b6e09c..de41105d 100644
--- a/src/certmonger.conf.5.in
+++ b/src/certmonger.conf.5.in
@@ -49,6 +49,7 @@ signing requests, and used when self\-signing certificates.
 @NO_MAN_DSA@\fIRSA\fR is supported.
 @MAN_DSA@\fIRSA\fR and \fIDSA\fR are supported.
 @MAN_EC@\fIEC\fR (also known as \fIECDSA\fR) is also supported.
+@MAN_ML_DSA@\fIML-DSA-44, ML-DSA-65, ML-DSA-87\fR are also supported.
 The default is \fIRSA\fP.
 
 .IP rsa_key_size
diff --git a/src/getcert.c b/src/getcert.c
index b4a02d4d..a8353285 100644
--- a/src/getcert.c
+++ b/src/getcert.c
@@ -894,6 +894,11 @@ request(const char *argv0, int argc, const char **argv)
 #ifdef CM_ENABLE_EC
 			    && (strcasecmp(poptarg, "ECDSA") != 0)
 			    && (strcasecmp(poptarg, "EC") != 0)
+#endif
+#ifdef CM_ENABLE_ML_DSA
+			    && (strcasecmp(poptarg, "ML-DSA-44") != 0)
+			    && (strcasecmp(poptarg, "ML-DSA-65") != 0)
+			    && (strcasecmp(poptarg, "ML-DSA-87") != 0)
 #endif
 			    ) {
 				printf(_("No support for generating \"%s\" keys.\n"),
@@ -905,6 +910,9 @@ request(const char *argv0, int argc, const char **argv)
 #endif
 #ifdef CM_ENABLE_EC
 				printf(" EC");
+#endif
+#ifdef CM_ENABLE_ML_DSA
+				printf(" ML-DSA-44 ML-DA-65 ML-DSA-87");
 #endif
 				printf("\n");
 				return 1;
@@ -2701,6 +2709,9 @@ rekey_or_resubmit(const char *argv0, const char *category, int argc,
 #ifdef CM_ENABLE_EC
 			    && (strcasecmp(poptarg, "ECDSA") != 0)
 			    && (strcasecmp(poptarg, "EC") != 0)
+#endif
+#ifdef CM_ENABLE_ML_DSA
+			    && (strcasecmp(poptarg, "ML-DSA-65") != 0)
 #endif
 			    ) {
 				printf(_("No support for generating \"%s\" keys.\n"),
diff --git a/src/keygen-n.c b/src/keygen-n.c
index 27c1efcc..3c3b6cd0 100644
--- a/src/keygen-n.c
+++ b/src/keygen-n.c
@@ -270,6 +270,14 @@ cm_keygen_n_main(int fd, struct cm_store_ca *ca, struct cm_store_entry *entry,
 		mech = CKM_EC_KEY_PAIR_GEN;
 		pmech = CKM_EC_KEY_PAIR_GEN;
 		break;
+#endif
+#ifdef CM_ENABLE_ML_DSA:
+	case cm_key_ml_dsa_44:
+	case cm_key_ml_dsa_65:
+	case cm_key_ml_dsa_87:
+		mech = CKM_ML_DSA_KEY_PAIR_GEN;
+		pmech = CKM_ML_DSA_KEY_PAIR_GEN;
+		break;
 #endif
 	default:
 		fprintf(status, "Unknown or unsupported key type.\n");
@@ -529,6 +537,13 @@ retry_gen:
 	case cm_key_ecdsa:
 		/* no parameters to generate */
 		break;
+#endif
+#ifdef CM_ENABLE_ML_DSA:
+	case cm_key_ml_dsa_44:
+	case cm_key_ml_dsa_65:
+	case cm_key_ml_dsa_87:
+		/* no parameters to generate */
+		break;
 #endif
 	default:
 		params = NULL;
@@ -564,6 +579,13 @@ retry_gen:
 				   &ecurve->oid, SEC_ObjectIDTemplate);
 		params = &ec_params;
 		break;
+#endif
+#ifdef CM_ENABLE_ML_DSA:
+	case cm_key_ml_dsa_44:
+	case cm_key_ml_dsa_65:
+	case cm_key_ml_dsa_87:
+		/* FIXME */
+		break;
 #endif
 	default:
 		params = NULL;
diff --git a/src/keygen-o.c b/src/keygen-o.c
index ff929fda..2f5a212b 100644
--- a/src/keygen-o.c
+++ b/src/keygen-o.c
@@ -111,10 +111,6 @@ cm_keygen_o_main(int fd, struct cm_store_ca *ca, struct cm_store_entry *entry,
 	int ret;
 	char *filename;
 	char *marker;
-#ifdef CM_ENABLE_EC
-	EC_KEY *ec;
-	int ecurve;
-#endif
 
 	status = fdopen(fd, "w");
 	if (status == NULL) {
@@ -294,6 +290,29 @@ retry_gen:
 		}
 		EVP_PKEY_CTX_free(pctx);
 		break;
+#endif
+#ifdef CM_ENABLE_ML_DSA
+	case cm_key_ml_dsa_44:
+	case cm_key_ml_dsa_65:
+	case cm_key_ml_dsa_87:
+		const char *alg = cm_store_algorithm_to_name(cm_key_algorithm);
+		EVP_PKEY_CTX *ctx = NULL;
+		if ((ctx = EVP_PKEY_CTX_new_from_name(NULL, alg, NULL)) == NULL) {
+			ERR_print_errors_fp(stderr);
+			cm_log(1, "Initializing key context failed\n");
+				_exit(CM_SUB_STATUS_INTERNAL_ERROR);
+		}
+		if (EVP_PKEY_keygen_init(ctx) <= 0) {
+			ERR_print_errors_fp(stderr);
+			cm_log(1, "keygen_init failed\n");
+			_exit(CM_SUB_STATUS_INTERNAL_ERROR);
+		}
+		if (EVP_PKEY_keygen(ctx, &pkey) <= 0) {
+			cm_log(1, "Error generating key.\n");
+			_exit(CM_SUB_STATUS_INTERNAL_ERROR);
+		}
+		EVP_PKEY_CTX_free(ctx);
+	break;
 #endif
 	default:
 		cm_log(1, "Unknown or unsupported key type.\n");
@@ -301,6 +320,9 @@ retry_gen:
 		break;
 	}
 
+	char *n = EVP_PKEY_get0_type_name(pkey);
+	cm_log(1, "debug: generated type %s for %d\n", n, cm_key_algorithm);
+
 	filename = strdup(entry->cm_key_storage_location);
 	marker = "";
 	keyfd = open(filename, O_RDWR | O_CREAT | O_EXCL, S_IRUSR | S_IWUSR);
diff --git a/src/keyiread-n.c b/src/keyiread-n.c
index c2f3928b..ed87b095 100644
--- a/src/keyiread-n.c
+++ b/src/keyiread-n.c
@@ -522,6 +522,7 @@ cm_keyiread_n_main(int fd, struct cm_store_ca *ca, struct cm_store_entry *entry,
 			cm_log(3, "Key is an EC key.\n");
 			alg = "EC";
 			break;
+		/* FIXME: case mlKey? */
 		case nullKey:
 		default:
 			cm_log(3, "Key is of an unknown type.\n");
@@ -581,6 +582,7 @@ cm_keyiread_n_main(int fd, struct cm_store_ca *ca, struct cm_store_entry *entry,
 					cm_log(3, "Next key is an EC key.\n");
 					alg = "EC";
 					break;
+				/* FIXME: case mlKey? */
 				case nullKey:
 				default:
 					cm_log(3, "Next key is of an unknown type.\n");
diff --git a/src/keyiread-o.c b/src/keyiread-o.c
index 51f7f829..383b2e7e 100644
--- a/src/keyiread-o.c
+++ b/src/keyiread-o.c
@@ -1,6 +1,6 @@
 /*
  * Copyright (C) 2009,2010,2011,2012,2014,2015,2017 Red Hat, Inc.
- * 
+ *
  * This program is free software: you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
  * the Free Software Foundation, either version 3 of the License, or
@@ -34,6 +34,17 @@
 #include <openssl/err.h>
 #include <openssl/pem.h>
 
+/************************
+ * FIXME: these definitions aren't public in openssl? wrong algos?
+************************/
+#define NID_ML_DSA_44          1457
+#define NID_ML_DSA_65          1458
+#define NID_ML_DSA_87          1459
+
+# define EVP_PKEY_ML_DSA_44           NID_ML_DSA_44
+# define EVP_PKEY_ML_DSA_65           NID_ML_DSA_65
+# define EVP_PKEY_ML_DSA_87           NID_ML_DSA_87
+
 #include <talloc.h>
 
 #include "keyiread.h"
@@ -152,6 +163,12 @@ cm_keyiread_o_main(int fd, struct cm_store_ca *ca, struct cm_store_entry *entry,
 		pubkey = "";
 		pubikey = "";
 		if (pkey != NULL) {
+			cm_log(1, "debug: pkey base is %d\n", util_EVP_PKEY_base_id(pkey));
+			cm_log(1, "debug: pkey ID is %d\n", EVP_PKEY_get_id(pkey));
+
+			char *n = EVP_PKEY_get0_type_name(pkey);
+			cm_log(1, "debug: type %s\n", n);
+
 			switch (util_EVP_PKEY_base_id(pkey)) {
 			case EVP_PKEY_RSA:
 				cm_log(3, "Key is an RSA key.\n");
@@ -168,6 +185,24 @@ cm_keyiread_o_main(int fd, struct cm_store_ca *ca, struct cm_store_entry *entry,
 				cm_log(3, "Key is an EC key.\n");
 				alg = "EC";
 				break;
+#endif
+#ifdef CM_ENABLE_ML_DSA
+			case 0:  /* no key type base */
+				char *name = EVP_PKEY_get0_type_name(pkey);
+				cm_log(3, "Key is a %s key.\n", name);
+
+				if (strcasecmp(name, "ML-DSA-44") == 0) {
+					alg = "ML-DSA-44";
+					break;
+				} else
+				if (strcasecmp(name, "ML-DSA-65") == 0) {
+					alg = "ML-DSA-65";
+					break;
+				} else
+				if (strcasecmp(name, "ML-DSA-87") == 0) {
+					alg = "ML-DSA-87";
+					break;
+				}
 #endif
 			default:
 				cm_log(3, "Key is for an unknown algorithm.\n");
@@ -190,6 +225,8 @@ cm_keyiread_o_main(int fd, struct cm_store_ca *ca, struct cm_store_entry *entry,
 					pubkey = cm_store_hex_from_bin(NULL, tmp, length);
 				}
 			}
+		} else {
+			cm_log(1, "pkey IS NULL\n");
 		}
 		fprintf(fp, "%s/%d/%s/%s\n", alg, bits, pubikey, pubkey);
 		if (nextpkey != NULL) {
@@ -209,6 +246,23 @@ cm_keyiread_o_main(int fd, struct cm_store_ca *ca, struct cm_store_entry *entry,
 				cm_log(3, "Next key is an EC key.\n");
 				alg = "EC";
 				break;
+#endif
+#ifdef CM_ENABLE_ML_DSA
+			case 0:  /* no key type base */
+				char *name = EVP_PKEY_get0_type_name(pkey);
+
+				if (strcasecmp(name, "ML-DSA-44") == 0) {
+					alg = "ML-DSA-44";
+					break;
+				} else
+				if (strcasecmp(name, "ML-DSA-65") == 0) {
+					alg = "ML-DSA-65";
+					break;
+				} else
+				if (strcasecmp(name, "ML-DSA-87") == 0) {
+					alg = "ML-DSA-87";
+					break;
+				}
 #endif
 			default:
 				cm_log(3, "Next key is for an unknown algorithm.\n");
diff --git a/src/keyiread.c b/src/keyiread.c
index db34c4b8..668b6d3e 100644
--- a/src/keyiread.c
+++ b/src/keyiread.c
@@ -140,9 +140,24 @@ cm_keyiread_read_data_from_buffer(struct cm_store_entry *entry, const char *p)
 	if (((q - p) == strlen("EC")) &&
 	    (strncasecmp(p, "EC", 2) == 0)) {
 		alg = cm_key_ecdsa;
+#endif
+#ifdef CM_ENABLE_ML_DSA
+	} else
+	if (((q - p) == strlen("ML-DSA-44")) &&
+	    (strncasecmp(p, "ML-DSA-44", 9) == 0)) {
+		alg = cm_key_ml_dsa_44;
+	} else
+	if (((q - p) == strlen("ML-DSA-65")) &&
+	    (strncasecmp(p, "ML-DSA-65", 9) == 0)) {
+		alg = cm_key_ml_dsa_65;
+	} else
+	if (((q - p) == strlen("ML-DSA-87")) &&
+	    (strncasecmp(p, "ML-DSA-87", 9) == 0)) {
+		alg = cm_key_ml_dsa_87;
 #endif
 	} else {
 		alg = cm_key_unspecified;
+		cm_log(1, "debug: determined unspcified from %d\n", alg);
 	}
 	if (alg != cm_key_unspecified) {
 		p = q + strspn(q, "/\r\n");
@@ -150,6 +165,7 @@ cm_keyiread_read_data_from_buffer(struct cm_store_entry *entry, const char *p)
 		if (p != q) {
 			size = atoi(p);
 			if (size > 0) {
+				cm_log(1, "debug: setting alg %d size %d\n", alg, size);
 				entry->cm_key_type.cm_key_algorithm = alg;
 				entry->cm_key_type.cm_key_size = size;
 				if (entry->cm_key_type.cm_key_gen_algorithm == 0) {
@@ -201,8 +217,23 @@ cm_keyiread_read_data_from_buffer(struct cm_store_entry *entry, const char *p)
 	if (((q - p) == strlen("EC")) &&
 	    (strncasecmp(p, "EC", 2) == 0)) {
 		alg = cm_key_ecdsa;
+#endif
+#ifdef CM_ENABLE_ML_DSA
+	} else
+	if (((q - p) == strlen("ML-DSA-44")) &&
+	    (strncasecmp(p, "ML-DSA-44", 9) == 0)) {
+		alg = cm_key_ml_dsa_44;
+	} else
+	if (((q - p) == strlen("ML-DSA-65")) &&
+	    (strncasecmp(p, "ML-DSA-65", 9) == 0)) {
+		alg = cm_key_ml_dsa_65;
+	} else
+	if (((q - p) == strlen("ML-DSA-87")) &&
+	    (strncasecmp(p, "ML-DSA-87", 9) == 0)) {
+		alg = cm_key_ml_dsa_87;
 #endif
 	} else {
+		cm_log(1, "debug: setting next unspcified from %s\n", p);
 		alg = cm_key_unspecified;
 	}
 	if (alg != cm_key_unspecified) {
diff --git a/src/prefs.c b/src/prefs.c
index 52ffc908..917a9f4e 100644
--- a/src/prefs.c
+++ b/src/prefs.c
@@ -515,6 +515,20 @@ cm_prefs_preferred_key_algorithm(void)
 			free(keytype);
 			return cm_key_rsa;
 		}
+#endif
+#ifdef CM_ENABLE_ML_DSA
+		else if (strcasecmp(keytype, "ML-DSA-44") == 0) {
+			free(keytype);
+			return cm_key_ml_dsa_44;
+		}
+		else if (strcasecmp(keytype, "ML-DSA-65") == 0) {
+			free(keytype);
+			return cm_key_ml_dsa_65;
+		}
+		else if (strcasecmp(keytype, "ML-DSA-87") == 0) {
+			free(keytype);
+			return cm_key_ml_dsa_87;
+		}
 #endif
 	}
 	return CM_DEFAULT_PUBKEY_TYPE;
diff --git a/src/store-files.c b/src/store-files.c
index 848de414..dee2fb0f 100644
--- a/src/store-files.c
+++ b/src/store-files.c
@@ -57,6 +57,7 @@
 
 static unsigned long long cm_entry_name_last, cm_ca_name_last;
 
+
 enum cm_store_file_field {
 	cm_store_file_field_invalid = 0,
 	cm_store_file_field_id,
@@ -840,6 +841,20 @@ cm_store_entry_read(void *parent, const char *filename, FILE *fp)
 				    (strcasecmp(s[i], "EC") == 0)) {
 					ret->cm_key_type.cm_key_algorithm =
 						cm_key_ecdsa;
+#endif
+#ifdef CM_ENABLE_ML_DSA
+				} else
+				if (strcasecmp(s[i], "ML-DSA-44") == 0) {
+					ret->cm_key_type.cm_key_algorithm =
+						cm_key_ml_dsa_44;
+				} else
+				if (strcasecmp(s[i], "ML-DSA-65") == 0) {
+					ret->cm_key_type.cm_key_algorithm =
+						cm_key_ml_dsa_65;
+				} else
+				if (strcasecmp(s[i], "ML-DSA-87") == 0) {
+					ret->cm_key_type.cm_key_algorithm =
+						cm_key_ml_dsa_87;
 #endif
 				} else {
 					ret->cm_key_type.cm_key_algorithm =
@@ -863,6 +878,20 @@ cm_store_entry_read(void *parent, const char *filename, FILE *fp)
 				    (strcasecmp(s[i], "EC") == 0)) {
 					ret->cm_key_type.cm_key_gen_algorithm =
 						cm_key_ecdsa;
+#endif
+#ifdef CM_ENABLE_ML_DSA
+				} else
+				if (strcasecmp(s[i], "ML-DSA-44") == 0) {
+					ret->cm_key_type.cm_key_gen_algorithm =
+						cm_key_ml_dsa_44;
+				} else
+				if (strcasecmp(s[i], "ML-DSA-65") == 0) {
+					ret->cm_key_type.cm_key_gen_algorithm =
+						cm_key_ml_dsa_65;
+				} else
+				if (strcasecmp(s[i], "ML-DSA-87") == 0) {
+					ret->cm_key_type.cm_key_gen_algorithm =
+						cm_key_ml_dsa_87;
 #endif
 				} else {
 					ret->cm_key_type.cm_key_gen_algorithm =
@@ -894,6 +923,20 @@ cm_store_entry_read(void *parent, const char *filename, FILE *fp)
 				    (strcasecmp(s[i], "EC") == 0)) {
 					ret->cm_key_next_type.cm_key_algorithm =
 						cm_key_ecdsa;
+#endif
+#ifdef CM_ENABLE_ML_DSA
+				} else
+				if (strcasecmp(s[i], "ML-DSA-44") == 0) {
+					ret->cm_key_type.cm_key_algorithm =
+						cm_key_ml_dsa_44;
+				} else
+				if (strcasecmp(s[i], "ML-DSA-65") == 0) {
+					ret->cm_key_type.cm_key_algorithm =
+						cm_key_ml_dsa_65;
+				} else
+				if (strcasecmp(s[i], "ML-DSA-87") == 0) {
+					ret->cm_key_type.cm_key_algorithm =
+						cm_key_ml_dsa_87;
 #endif
 				} else {
 					ret->cm_key_next_type.cm_key_algorithm =
@@ -917,6 +960,20 @@ cm_store_entry_read(void *parent, const char *filename, FILE *fp)
 				    (strcasecmp(s[i], "EC") == 0)) {
 					ret->cm_key_next_type.cm_key_gen_algorithm =
 						cm_key_ecdsa;
+#endif
+#ifdef CM_ENABLE_ML_DSA
+				} else
+				if (strcasecmp(s[i], "ML-DSA-44") == 0) {
+					ret->cm_key_type.cm_key_gen_algorithm =
+						cm_key_ml_dsa_44;
+				} else
+				if (strcasecmp(s[i], "ML-DSA-65") == 0) {
+					ret->cm_key_type.cm_key_gen_algorithm =
+						cm_key_ml_dsa_65;
+				} else
+				if (strcasecmp(s[i], "ML-DSA-87") == 0) {
+					ret->cm_key_type.cm_key_gen_algorithm =
+						cm_key_ml_dsa_87;
 #endif
 				} else {
 					ret->cm_key_next_type.cm_key_gen_algorithm =
@@ -1829,6 +1886,20 @@ cm_store_entry_write(FILE *fp, struct cm_store_entry *entry)
 		cm_store_file_write_str(fp, cm_store_entry_field_key_type,
 					"EC");
 		break;
+#endif
+#ifdef CM_ENABLE_ML_DSA
+	case cm_key_ml_dsa_44:
+		cm_store_file_write_str(fp, cm_store_entry_field_key_type,
+					"ML-DSA-44");
+		break;
+	case cm_key_ml_dsa_65:
+		cm_store_file_write_str(fp, cm_store_entry_field_key_type,
+					"ML-DSA-65");
+		break;
+	case cm_key_ml_dsa_87:
+		cm_store_file_write_str(fp, cm_store_entry_field_key_type,
+					"ML-DSA-87");
+		break;
 #endif
 	}
 	switch (entry->cm_key_type.cm_key_gen_algorithm) {
@@ -1851,6 +1922,20 @@ cm_store_entry_write(FILE *fp, struct cm_store_entry *entry)
 		cm_store_file_write_str(fp, cm_store_entry_field_key_gen_type,
 					"EC");
 		break;
+#endif
+#ifdef CM_ENABLE_ML_DSA
+	case cm_key_ml_dsa_44:
+		cm_store_file_write_str(fp, cm_store_entry_field_key_gen_type,
+					"ML-DSA-44");
+		break;
+	case cm_key_ml_dsa_65:
+		cm_store_file_write_str(fp, cm_store_entry_field_key_gen_type,
+					"ML-DSA-65");
+		break;
+	case cm_key_ml_dsa_87:
+		cm_store_file_write_str(fp, cm_store_entry_field_key_gen_type,
+					"ML-DSA-87");
+		break;
 #endif
 	}
 	cm_store_file_write_int(fp, cm_store_entry_field_key_size,
@@ -1877,6 +1962,20 @@ cm_store_entry_write(FILE *fp, struct cm_store_entry *entry)
 		cm_store_file_write_str(fp, cm_store_entry_field_key_next_type,
 					"EC");
 		break;
+#endif
+#ifdef CM_ENABLE_ML_DSA
+	case cm_key_ml_dsa_44:
+		cm_store_file_write_str(fp, cm_store_entry_field_key_next_type,
+					"ML-DSA-44");
+		break;
+	case cm_key_ml_dsa_65:
+		cm_store_file_write_str(fp, cm_store_entry_field_key_next_type,
+					"ML-DSA-65");
+		break;
+	case cm_key_ml_dsa_87:
+		cm_store_file_write_str(fp, cm_store_entry_field_key_next_type,
+					"ML-DSA-87");
+		break;
 #endif
 	}
 	switch (entry->cm_key_next_type.cm_key_gen_algorithm) {
@@ -1899,6 +1998,20 @@ cm_store_entry_write(FILE *fp, struct cm_store_entry *entry)
 		cm_store_file_write_str(fp, cm_store_entry_field_key_next_gen_type,
 					"EC");
 		break;
+#endif
+#ifdef CM_ENABLE_ML_DSA
+	case cm_key_ml_dsa_44:
+		cm_store_file_write_str(fp, cm_store_entry_field_key_next_gen_type,
+					"ML-DSA-44");
+		break;
+	case cm_key_ml_dsa_65:
+		cm_store_file_write_str(fp, cm_store_entry_field_key_next_gen_type,
+					"ML-DSA-65");
+		break;
+	case cm_key_ml_dsa_87:
+		cm_store_file_write_str(fp, cm_store_entry_field_key_next_gen_type,
+					"ML-DSA-87");
+		break;
 #endif
 	}
 	cm_store_file_write_int(fp, cm_store_entry_field_key_next_size,
@@ -2989,3 +3102,37 @@ cm_store_ca_dup(void *parent, struct cm_store_ca *ca)
 
 	return ret;
 }
+
+char *cm_store_algorithm_to_name(int algorithm)
+{
+	/* See cm_key_algorithm in store-int.h */
+	switch(algorithm) {
+	case cm_key_rsa:
+		return "RSA";
+		break;
+#ifdef CM_ENABLE_DSA
+	case cm_key_dsa:
+		return "DSA";
+		break;
+#endif
+#ifdef CM_ENABLE_EC
+	case cm_key_ecdsa:
+		return "EC";
+		break;
+#endif
+#ifdef CM_ENABLE_ML_DSA
+	case cm_key_ml_dsa_44:
+		return "ML-DSA-44";
+		break;
+	case cm_key_ml_dsa_65:
+		return "ML-DSA-65";
+		break;
+	case cm_key_ml_dsa_87:
+		return "ML-DSA-87";
+		break;
+#endif
+	default:
+		return "UNKNOWN";
+		break;
+	}
+};
diff --git a/src/store-int.h b/src/store-int.h
index 7557919a..fca98776 100644
--- a/src/store-int.h
+++ b/src/store-int.h
@@ -39,6 +39,11 @@ struct cm_store_entry {
 #endif
 #ifdef CM_ENABLE_EC
 			cm_key_ecdsa,
+#endif
+#ifdef CM_ENABLE_ML_DSA
+			cm_key_ml_dsa_44,
+			cm_key_ml_dsa_65,
+			cm_key_ml_dsa_87
 #endif
 		} cm_key_algorithm, cm_key_gen_algorithm;
 		int cm_key_size, cm_key_gen_size;
diff --git a/src/store.h b/src/store.h
index 2c2b3267..67814117 100644
--- a/src/store.h
+++ b/src/store.h
@@ -70,5 +70,6 @@ void cm_store_set_if_not_set_as(void *parent, char ***dest, char **src);
 int cm_store_utf8_to_bmp_string(char *s,
 				unsigned char **bmp, unsigned int *len);
 char *cm_store_utf8_from_bmp_string(unsigned char *bmp, unsigned int len);
+char *cm_store_algorithm_to_name(int algorithm);
 
 #endif
diff --git a/src/submit-e.c b/src/submit-e.c
index 69b4f8e2..8f3c087d 100644
--- a/src/submit-e.c
+++ b/src/submit-e.c
@@ -883,6 +883,17 @@ cm_submit_e_helper_main(int fd, struct cm_store_ca *ca,
 	case cm_key_ecdsa:
 		key_type = "EC";
 		break;
+#endif
+#ifdef CM_ENABLE_ML_DSA:
+	case cm_key_ml_dsa_44:
+		key_type = "ML-DSA-44";
+		break;
+	case cm_key_ml_dsa_65:
+		key_type = "ML-DSA-65";
+		break;
+	case cm_key_ml_dsa_87:
+		key_type = "ML-DSA-87";
+		break;
 #endif
 	case cm_key_unspecified:
 		key_type = NULL;
diff --git a/src/tdbush.c b/src/tdbush.c
index d4408632..bb83a388 100644
--- a/src/tdbush.c
+++ b/src/tdbush.c
@@ -1316,6 +1316,20 @@ base_add_request(DBusConnection *conn, DBusMessage *msg,
 		    (strcasecmp(param->value.s, "EC") == 0)) {
 			new_entry->cm_key_type.cm_key_gen_algorithm =
 				cm_key_ecdsa;
+#endif
+#ifdef CM_ENABLE_ML_DSA
+		} else
+		if (strcasecmp(param->value.s, "ML-DSA-44") == 0) {
+			new_entry->cm_key_type.cm_key_gen_algorithm =
+				cm_key_ml_dsa_44;
+		} else
+		if (strcasecmp(param->value.s, "ML-DSA-65") == 0) {
+			new_entry->cm_key_type.cm_key_gen_algorithm =
+				cm_key_ml_dsa_65;
+		} else
+		if (strcasecmp(param->value.s, "ML-DSA-87") == 0) {
+			new_entry->cm_key_type.cm_key_gen_algorithm =
+				cm_key_ml_dsa_87;
 #endif
 		} else {
 			cm_log(1, "No support for generating \"%s\" keys.\n",
@@ -1361,6 +1375,16 @@ base_add_request(DBusConnection *conn, DBusMessage *msg,
 			new_entry->cm_key_type.cm_key_gen_size = CM_MINIMUM_EC_KEY_SIZE;
 		}
 		break;
+#endif
+#ifdef CM_ENABLE_ML_DSA:
+	case cm_key_ml_dsa_44:
+	case cm_key_ml_dsa_65:
+	case cm_key_ml_dsa_87:
+		if (new_entry->cm_key_type.cm_key_gen_size < CM_MINIMUM_ML_DSA_KEY_SIZE)
+		{
+			new_entry->cm_key_type.cm_key_gen_size = CM_MINIMUM_ML_DSA_KEY_SIZE;
+		}
+		break;
 #endif
 	case cm_key_unspecified:
 	default:
@@ -1837,6 +1861,11 @@ base_get_supported_key_types(DBusConnection *conn, DBusMessage *msg,
 #endif
 #ifdef CM_ENABLE_EC
 		"EC",
+#endif
+#ifdef CM_ENABLE_ML_DSA
+		"ML-DSA-44",
+		"ML-DSA-65",
+		"ML-DSA-87",
 #endif
 		NULL
 	};
@@ -2924,27 +2953,7 @@ request_get_key_type_and_size(DBusConnection *conn, DBusMessage *msg,
 		return DBUS_HANDLER_RESULT_NOT_YET_HANDLED;
 	}
 	rep = dbus_message_new_method_return(msg);
-	switch (entry->cm_key_type.cm_key_algorithm) {
-	case cm_key_unspecified:
-		type = "UNKNOWN";
-		break;
-	case cm_key_rsa:
-		type = "RSA";
-		break;
-#ifdef CM_ENABLE_DSA
-	case cm_key_dsa:
-		type = "DSA";
-		break;
-#endif
-#ifdef CM_ENABLE_EC
-	case cm_key_ecdsa:
-		type = "EC";
-		break;
-#endif
-	default:
-		type = "UNKNOWN";
-		break;
-	}
+	type = cm_store_algorithm_to_name(entry->cm_key_type.cm_key_algorithm);
 	if (rep != NULL) {
 		size = entry->cm_key_type.cm_key_size;
 		cm_tdbusm_set_sn(rep, type, size);
@@ -3250,6 +3259,17 @@ request_modify(DBusConnection *conn, DBusMessage *msg,
 				if ((strcasecmp(param->value.s, "ECDSA") == 0) ||
 				    (strcasecmp(param->value.s, "EC") == 0)) {
 					entry->cm_key_type.cm_key_gen_algorithm = cm_key_ecdsa;
+#endif
+#ifdef CM_ENABLE_ML_DSA
+				} else
+				if (strcasecmp(param->value.s, "ML-DSA-44") == 0) {
+					entry->cm_key_type.cm_key_gen_algorithm = cm_key_ml_dsa_44;
+				} else
+				if (strcasecmp(param->value.s, "ML-DSA-65") == 0) {
+					entry->cm_key_type.cm_key_gen_algorithm = cm_key_ml_dsa_65;
+				} else
+				if (strcasecmp(param->value.s, "ML-DSA-87") == 0) {
+					entry->cm_key_type.cm_key_gen_algorithm = cm_key_ml_dsa_87;
 #endif
 				} else {
 					cm_log(1, "No support for generating \"%s\" keys.\n",
@@ -4048,6 +4068,17 @@ request_prop_get_key_type(struct cm_context *ctx, void *parent,
 	case cm_key_ecdsa:
 		return "EC";
 		break;
+#endif
+#ifdef CM_ENABLE_ML_DSA:
+       case cm_key_ml_dsa_44:
+		return "ML-DSA-44";
+		break;
+       case cm_key_ml_dsa_65:
+		return "ML-DSA-65";
+		break;
+       case cm_key_ml_dsa_87:
+		return "ML-DSA-87";
+		break;
 #endif
 	}
 	return "";
@@ -4070,6 +4101,12 @@ request_prop_get_key_size(struct cm_context *ctx, void *parent,
 #endif
 #ifdef CM_ENABLE_EC
 	case cm_key_ecdsa:
+#endif
+#ifdef CM_ENABLE_ML_DSA:
+       case cm_key_ml_dsa_44:
+       case cm_key_ml_dsa_65:
+       case cm_key_ml_dsa_87:
+		/* fall through */
 #endif
 		return entry->cm_key_type.cm_key_size;
 		break;
diff --git a/tests/002-keygen-ml-dsa/expected.out b/tests/002-keygen-ml-dsa/expected.out
new file mode 100644
index 00000000..b3039590
--- /dev/null
+++ b/tests/002-keygen-ml-dsa/expected.out
@@ -0,0 +1,14 @@
+[openssl:44]
+OK.
+OK (ML-DSA-44:10496).
+[openssl:65]
+OK.
+OK (ML-DSA-65:15616).
+[openssl:87]
+OK.
+OK (ML-DSA-87:20736).
+[openssl:rosubdir]
+Failed to save FILE:${tmpdir}/rosubdir/sample.87: need fs permissions.
+[openssl:rwsubdir]
+Failed to save FILE:${tmpdir}/rwsubdir/sample.87: need fs permissions.
+Test complete.
diff --git a/tests/002-keygen-ml-dsa/run.sh b/tests/002-keygen-ml-dsa/run.sh
new file mode 100755
index 00000000..55360f7d
--- /dev/null
+++ b/tests/002-keygen-ml-dsa/run.sh
@@ -0,0 +1,43 @@
+#!/bin/bash -e
+
+cd "$tmpdir"
+
+source "$srcdir"/functions
+initnssdb "$tmpdir"
+
+for size in 44 65 87 ; do
+	echo "[openssl:$size]"
+	# Generate a key.
+	cat > entry.$size <<- EOF
+	key_storage_type=FILE
+	key_storage_location=$tmpdir/sample.$size
+	key_gen_size=$size
+	key_gen_type=ML-DSA-$size
+	EOF
+	$toolsdir/keygen entry.$size
+	# Read the size.
+	sed -i 's,^key_gen_size.*,,g' entry.$size
+	$toolsdir/keyiread entry.$size
+done
+
+echo "[openssl:rosubdir]"
+cat > entry.$size <<- EOF
+key_storage_type=FILE
+key_storage_location=$tmpdir/rosubdir/sample.$size
+key_gen_size=$size
+key_gen_type=ML-DSA-$size
+EOF
+$toolsdir/keygen entry.$size || true
+
+echo "[openssl:rwsubdir]"
+cat > entry.$size <<- EOF
+key_storage_type=FILE
+key_storage_location=$tmpdir/rwsubdir/sample.$size
+key_gen_size=$size
+key_gen_type=ML-DSA-$size
+EOF
+touch $tmpdir/rwsubdir/sample.$size
+chmod u-w $tmpdir/rwsubdir/sample.$size
+$toolsdir/keygen entry.$size || true
+
+echo Test complete.
diff --git a/tests/028-dbus/expected.out b/tests/028-dbus/expected.out
index 86cba020..28bed7c7 100644
--- a/tests/028-dbus/expected.out
+++ b/tests/028-dbus/expected.out
@@ -185,7 +185,7 @@ dbus.Array([dbus.ObjectPath('/org/fedorahosted/certmonger/cas/CA1'), dbus.Object
 dbus.Array([dbus.ObjectPath('/org/fedorahosted/certmonger/requests/Request2')], signature=dbus.Signature('o'))
 
 [ /org/fedorahosted/certmonger: org.fedorahosted.certmonger.get_supported_key_types ]
-dbus.Array([dbus.String('RSA'), dbus.String('DSA'), dbus.String('EC')], signature=dbus.Signature('s'))
+dbus.Array([dbus.String('RSA'), dbus.String('DSA'), dbus.String('EC'), dbus.String('ML-DSA-44'), dbus.String('ML-DSA-65'), dbus.String('ML-DSA-87')], signature=dbus.Signature('s'))
 
 [ /org/fedorahosted/certmonger: org.fedorahosted.certmonger.get_supported_key_storage ]
 dbus.Array([dbus.String('NSSDB'), dbus.String('FILE')], signature=dbus.Signature('s'))
@@ -712,7 +712,7 @@ dbus.Array([], signature=dbus.Signature('s'))
 </node>
 
 [ /org/fedorahosted/certmonger/cas/CA2: org.fedorahosted.certmonger.ca.get_config_file_path ]
-$tmpdir/cas/20180327134236
+$tmpdir/cas/20250807195047
 
 [ /org/fedorahosted/certmonger/cas/CA2: org.fedorahosted.certmonger.ca.get_nickname ]
 SelfSign
@@ -827,7 +827,7 @@ dbus.Array([], signature=dbus.Signature('s'))
 </node>
 
 [ /org/fedorahosted/certmonger/cas/CA3: org.fedorahosted.certmonger.ca.get_config_file_path ]
-$tmpdir/cas/20180327134236-1
+$tmpdir/cas/20250807195047-1
 
 [ /org/fedorahosted/certmonger/cas/CA3: org.fedorahosted.certmonger.ca.get_nickname ]
 IPA
@@ -940,7 +940,7 @@ dbus.Array([], signature=dbus.Signature('s'))
 </node>
 
 [ /org/fedorahosted/certmonger/cas/CA4: org.fedorahosted.certmonger.ca.get_config_file_path ]
-$tmpdir/cas/20180327134236-3
+$tmpdir/cas/20250807195047-2
 
 [ /org/fedorahosted/certmonger/cas/CA4: org.fedorahosted.certmonger.ca.get_nickname ]
 dogtag-ipa-renew-agent
diff --git a/tests/tools/keyiread.c b/tests/tools/keyiread.c
index 5934124a..f6140ca9 100644
--- a/tests/tools/keyiread.c
+++ b/tests/tools/keyiread.c
@@ -63,6 +63,17 @@ type_name(enum cm_key_algorithm alg)
 	case cm_key_ecdsa:
 		return "EC";
 		break;
+#endif
+#ifdef CM_ENABLE_ML_DSA:
+	case cm_key_ml_dsa_44:
+		return "ML-DSA-44";
+		break;
+	case cm_key_ml_dsa_65:
+		return "ML-DSA-65";
+		break;
+	case cm_key_ml_dsa_87:
+		return "ML-DSA-87";
+		break;
 #endif
 	default:
 		assert(0);
@@ -85,6 +96,12 @@ munge_key_size(struct cm_key_type *key)
 #ifdef CM_ENABLE_EC
 	case cm_key_ecdsa:
 		break;
+#endif
+#ifdef CM_ENABLE_ML_DSA:
+	case cm_key_ml_dsa_44:
+	case cm_key_ml_dsa_65:
+	case cm_key_ml_dsa_87:
+		break;
 #endif
 	default:
 		assert(0);
-- 
2.52.0

